modules/protocol/src/vector.ts

```diff
@@ -263,50 +257,117 @@ export class Vector implements IVectorProtocol {
-        //   this.logger.info({ channelAddress: updatedChannel.channelAddress }, "Registering channel for dispute events");
-        //   const registrationRes = await this.chainReader.registerChannel(
-        //     updatedChannel.channelAddress,
-        //     updatedChannel.networkContext.chainId,
-        //   );
-        //   if (registrationRes.isError) {
-        //     this.logger.warn(
-        //       { ...jsonifyError(registrationRes.getError()!) },
-        //       "Failed to register channel for dispute watching",
-        //     );
-        //   }
-        // }
+        const { updatedChannel, updatedActiveTransfers, updatedTransfer } = result.getValue();
 
         this.evts[ProtocolEventName.CHANNEL_UPDATE_EVENT].post({
           updatedChannelState: updatedChannel,
           updatedTransfers: updatedActiveTransfers,
           updatedTransfer,
         });
-        this.logger.debug({ method, methodId }, "Method complete");
+        this.logger.debug({ method, methodId, channelNonce: updatedChannel.nonce }, "Applied received update");
+
+        // Check if you are currently proposing an update
+        const proposed = this.inProgressUpdate[received.update.channelAddress];
+
+        // If no, return
+        if (!proposed) {
+          return;
+        }
+
+        // If so, check if the update includes our proposed update
+        const proposedTransfer =
+          proposed.type === UpdateType.create ? getTransferFromUpdate(proposed, updatedChannel) : undefined;
+        const included = this.includesOurProposedUpdate(
+          received.update,
+          proposed,
+          // TODO: make it *always* return updatedActiveTransfers
+          updatedActiveTransfers!,
+          proposedTransfer,
+        );
+
+        // If it does include our proposed update, add inbound to the
+        // queue and remove currently processing update
+        if (!included.isError && included.getValue()) {
+          // Remove the proposed update from our tracker
+          this.inProgressUpdate[received.update.channelAddress] = undefined;
+          this.logger.debug(
+            { method, methodId, channelNonce: updatedChannel.nonce },
+            "Proposed update included, not regenerating",
+          );
+          // TODO: resume queue without inserting update
+          this.logger.debug({ method, methodId }, "Method complete");
+          return;
+        }
+
+        // There is a case here where included is an error, in which
+        // case you should retry the proposed update anyway (errors
+        // should fall through to validation)
+
+        // If it does not, insert previously in progress update
+        // at the front of queue
+        const regenerated = getParamsFromUpdate(proposed);
+        if (regenerated.isError) {
+          return;
+        }
+        const processedResult = await this.addToFrontOfQueue(regenerated.getValue(), received.update);
```

Comment:
This has me a little worried that it's not just taking the same path. (Eg: At the top of the method this uses `addToQueue` instead of `addToFrontOfQueue`. The proposed update is just the same now as a normal update that would have been added later, so a more natural way to do this in a lock-free system would be to put a `while (true)` at the top of this method so that everything retries including nonce setting, etc. `while(true)` is a very common pattern to see in a lock-free construction.

In addition to the above, I find a lot of this hard to read but am hoping that implementing the queue or something to serialize access to data will tame this. I've got a pretty good idea how to approach the queue without the sync protocol and disputes... but a queue without those isn't good enough.

Getting this down to "obviously correct" is going to be a struggle but that might be part of the territory.